bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -1655,37 +1655,53 @@ impl<'a> CircuitInputStateRef<'a> {
         dst_addr: impl Into<MemoryAddress>,
         src_addr_end: impl Into<MemoryAddress>,
         bytes_left: impl Into<MemoryAddress>,
-        memory_updated: &Memory,
     ) -> Result<(CopyEventSteps, CopyEventPrevBytes), Error> {
-        let src_addr = src_addr.into();
-        let dst_addr = dst_addr.into();
-        let src_addr_end = src_addr_end.into();
-        let bytes_left = bytes_left.into();
+        let src_addr = src_addr.into().0;
+        let dst_addr = dst_addr.into().0;
+        let src_addr_end = src_addr_end.into().0;
+        let bytes_left = bytes_left.into().0;
+        let src_copy_end = src_addr + bytes_left;
+        let dst_copy_end = dst_addr + bytes_left;
 
         if bytes_left == 0 {
             return Ok((vec![], vec![]));
         }
 
+        // Extend call memory.
+        let memory = &mut self.call_ctx_mut()?.memory;
+        memory.extend_for_range(dst_addr.into(), bytes_left.into());
+
+        // Align memory range.
         let dst_range = MemoryWordRange::align_range(dst_addr, bytes_left);
-        let code_slot_bytes = memory_updated.read_chunk(dst_range);
+        let dst_begin_slot = dst_range.start_slot().0;
+        let dst_end_slot = dst_range.end_slot().0.min(memory.len());
+        assert!(dst_begin_slot <= dst_addr && dst_end_slot >= dst_copy_end);
+
+        // Combine the slot bytes.
+        let bytes_to_copy = bytecode.code[src_addr..src_addr_end.min(src_copy_end)]
+            .iter()
+            .map(|b| b.value);
+        let padding_bytes =
+            repeat(0).take(src_copy_end.checked_sub(src_addr_end).unwrap_or_default());
+        let slot_bytes: Vec<u8> = memory[dst_begin_slot..dst_addr]
+            .iter()
+            .cloned()
+            .chain(bytes_to_copy)
+            .chain(padding_bytes)
+            .chain(memory[dst_copy_end..dst_end_slot].iter().cloned())
+            .collect();
 
-        let dst_begin_slot = dst_range.start_slot();
         let copy_steps = CopyEventStepsBuilder::new()
-            .source(&bytecode.code[..src_addr_end.0])
+            .source(&bytecode.code[..src_addr_end])
```

Comment:
Throughout the codebase src_addr_end really means src_addr_*limit* and we should not assume that it matches the real source end. Sometimes it is greater.

I recommend to put src_addr_end.min(src_copy_end) into a variable and use it instead.

Reply:
Fix to set `src_addr_end = src_addr_end.min(src_copy_end)` in [9b70103](https://github.com/scroll-tech/zkevm-circuits/pull/629/commits/9b70103d77c1ea7c22d6ea2e48768368417c78b8), and it makes the below code clear.